src/Adapter/DoctrineItems.php

```diff
@@ -0,0 +1,354 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Libero\ContentApiBundle\Adapter;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\FetchMode;
+use Doctrine\DBAL\LockMode;
+use Doctrine\DBAL\ParameterType;
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\DBAL\Types\Type;
+use IteratorAggregate;
+use Libero\ContentApiBundle\Exception\ItemNotFound;
+use Libero\ContentApiBundle\Exception\UnexpectedVersionNumber;
+use Libero\ContentApiBundle\Exception\VersionNotFound;
+use Libero\ContentApiBundle\Model\ItemId;
+use Libero\ContentApiBundle\Model\ItemListPage;
+use Libero\ContentApiBundle\Model\Items;
+use Libero\ContentApiBundle\Model\ItemVersion;
+use Libero\ContentApiBundle\Model\ItemVersionNumber;
+use Traversable;
+use function array_column;
+use function array_map;
+use function array_pop;
+use function count;
+use function rewind;
+use function sprintf;
+
+final class DoctrineItems implements IteratorAggregate, Items
+{
+    private $connection;
+    private $tableItems;
+    private $tableVersions;
+
+    public function __construct(Connection $connection, string $tablePrefix)
+    {
+        $this->connection = $connection;
+        $this->tableItems = "{$tablePrefix}_items";
+        $this->tableVersions = "{$tablePrefix}_versions";
+    }
+
+    public function add(ItemVersion $item) : void
+    {
+        $this->connection->transactional(
+            function () use ($item) : void {
+                try {
+                    $current = $this->getCurrentVersion($item->getId(), LockMode::PESSIMISTIC_WRITE);
+                } catch (ItemNotFound $e) {
+                    if (1 !== $item->getVersion()->toInt()) {
```

Comment:
could probably move this assertion into `insertNewItem` to ensure it cannot be used in the wrong conditions?